File: code/java/2402.java
// https://leetcode.com/problems/meeting-rooms-iii/
class Solution {
    public int mostBooked(int n, int[][] meetings) {
        long[] availableTime = new long[n];
        int[] freq = new int[n];
        Arrays.sort(meetings, (a, b) -> Integer.compare(a[0], b[0]));
        for (int[] meeting : meetings) {
            int start = meeting[0], end = meeting[1];
            long minAvailableTime = Long.MAX_VALUE;
            int minRoom = 0;
            boolean foundRoom = false;
            for (int i = 0; i < n; i++) {
                if (availableTime[i] <= start) {
                    availableTime[i] = end;
                    freq[i]++;
                    foundRoom = true;
                    break;
                }
                if (availableTime[i] < minAvailableTime) {
                    minAvailableTime = availableTime[i];
                    minRoom = i;
                }
            }
            if (!foundRoom) {
                availableTime[minRoom] += end - start;
                freq[minRoom]++;
            }
        }
        int res = 0;
        for (int i = 0; i < n; i++)
            if (freq[i] > freq[res])
                res = i;
        return res;
    }
}

File: code/python/2402.py
# https://leetcode.com/problems/meeting-rooms-iii/
class Solution:
    def mostBooked(self, n: int, meetings: List[List[int]]) -> int:
        free, busy = list(range(n)), []
        heapq.heapify(free)
        meetings.sort(key=lambda x: x[0])
        freq = [0] * n
        for m_start, m_end in meetings:
            while busy and busy[0][0] <= m_start:
                heapq.heappush(free, heapq.heappop(busy)[1])
            if free:
                room = heapq.heappop(free)
            else:
                now, room = heapq.heappop(busy)
                m_end += now - m_start
            heapq.heappush(busy, (m_end, room))
            freq[room] += 1
        return max(range(n), key=lambda x: (freq[x], -x))
